cache/lru_secondary_cache.cc

```diff
@@ -0,0 +1,159 @@
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "cache/lru_secondary_cache.h"
+
+#include <memory>
+
+#include "util/compression.h"
+#include "util/string_util.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+LRUSecondaryCache::LRUSecondaryCache(
+    size_t capacity, int num_shard_bits, bool strict_capacity_limit,
+    double high_pri_pool_ratio,
+    std::shared_ptr<MemoryAllocator> memory_allocator, bool use_adaptive_mutex,
+    CacheMetadataChargePolicy metadata_charge_policy,
+    CompressionType compression_type, uint32_t compress_format_version)
+    : compression_type_(compression_type),
+      compress_format_version_(compress_format_version) {
+  cache_ = NewLRUCache(capacity, num_shard_bits, strict_capacity_limit,
+                       high_pri_pool_ratio, memory_allocator,
+                       use_adaptive_mutex, metadata_charge_policy);
+}
+
+LRUSecondaryCache::~LRUSecondaryCache() { cache_.reset(); }
+
+std::unique_ptr<SecondaryCacheResultHandle> LRUSecondaryCache::Lookup(
+    const Slice& key, const Cache::CreateCallback& create_cb, bool /*wait*/) {
+  std::unique_ptr<SecondaryCacheResultHandle> handle;
+  Cache::Handle* lru_handle = cache_->Lookup(key);
+  if (lru_handle == nullptr) {
+    return handle;
+  }
+
+  void* value = nullptr;
+  size_t charge = 0;
+  Status s;
+
+  if (compression_type_ == kNoCompression) {
+    s = create_cb(cache_->Value(lru_handle), cache_->GetCharge(lru_handle),
+                  &value, &charge);
+  } else {
+    UncompressionContext uncompression_context(compression_type_);
+    UncompressionInfo uncompression_info(uncompression_context,
+                                         UncompressionDict::GetEmptyDict(),
+                                         compression_type_);
+
+    size_t uncompressed_size = 0;
+    CacheAllocationPtr uncompressed;
+    uncompressed =
+        UncompressData(uncompression_info, (char*)cache_->Value(lru_handle),
+                       cache_->GetCharge(lru_handle), &uncompressed_size,
+                       compress_format_version_);
+
+    if (!uncompressed) {
+      return handle;
+    }
+
+    s = create_cb(uncompressed.get(), uncompressed_size, &value, &charge);
+  }
+
+  if (!s.ok()) {
+    return handle;
+  }
+  handle.reset(new LRUSecondaryCacheResultHandle(cache_.get(), lru_handle,
+                                                 value, charge));
+  return handle;
+}
+
+Status LRUSecondaryCache::Insert(const Slice& key, void* value,
+                                 const Cache::CacheItemHelper* helper) {
+  size_t size = (*helper->size_cb)(value);
+  char* buf = new char[size];
+  Status s = (*helper->saveto_cb)(value, 0, size, buf);
+  if (!s.ok()) {
+    delete[] buf;
+    return s;
+  }
+  Slice val(buf, size);
+
+  if (compression_type_ != kNoCompression) {
+    CompressionOptions compression_opts;
+    CompressionContext compression_context(compression_type_);
+    uint64_t sample_for_compression = 1;
```

Comment:
I know its moot, since sampling is only done in BlockBasedTableBuilder, but I suggest setting sample_for_compression to 0 to avoid any confusion for readers.